esp-metadata/src/lib.rs

```diff
@@ -562,22 +566,34 @@ impl Config {
                 .iter()
                 .any(|p| peri.name.eq_ignore_ascii_case(p))
             {
+                all_peripherals.push(quote::quote! { stable #tokens });
                 stable.push(tokens);
             } else {
+                all_peripherals.push(quote::quote! { unstable #tokens });
                 unstable.push(tokens);
             }
         }
 
+        let for_each = generate_for_each_macro("peripheral", &all_peripherals);
+
         quote::quote! {
-            crate::peripherals! {
-                peripherals: [
-                    #(#stable,)*
-                ],
-                unstable_peripherals: [
-                    #(#unstable,)*
-                ],
-                pins: [#(#gpios,)*]
+            #for_each
+
+            // Creates the Peripherals struct
+            macro_rules! define_peripherals {
```

Comment:
This is still not great, I actually need a macro that exposes the list of peripherals. The `for_each_peripheral` almost does this, but it calls a single macro N times, instead of repetition, which means it's not suitable for defining a struct. This isn't hard to solve for a single case, but it makes me wonder if we can do better, generally.